File: app-examples/example-jv-02threading/src/androidTest/java/foo/bar/example/forethreading/ui/StateBuilder.java
package foo.bar.example.forethreading.ui;

import androidx.test.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;
import co.early.fore.core.WorkMode;
import foo.bar.example.forethreading.App;
import foo.bar.example.forethreading.OG;
import foo.bar.example.forethreading.feature.counter.CounterWithLambdas;
import foo.bar.example.forethreading.feature.counter.CounterWithProgress;

import static org.mockito.Mockito.when;

/**
 *
 */
public class StateBuilder {

    private CounterWithLambdas mockCounterWithLambdas;
    private CounterWithProgress mockCounterWithProgress;

    StateBuilder(CounterWithLambdas mockCounterWithLambdas, CounterWithProgress mockCounterWithProgress) {
        this.mockCounterWithLambdas = mockCounterWithLambdas;
        this.mockCounterWithProgress = mockCounterWithProgress;
    }

    StateBuilder counterBasicIsBusy(boolean busy) {
        when(mockCounterWithLambdas.isBusy()).thenReturn(busy);
        return this;
    }

    StateBuilder counterWithProgressIsBusy(boolean busy) {
        when(mockCounterWithProgress.isBusy()).thenReturn(busy);
        return this;
    }

    StateBuilder counterBasicCount(int count) {
        when(mockCounterWithLambdas.getCount()).thenReturn(count);
        return this;
    }

    StateBuilder counterWithProgressCount(int count) {
        when(mockCounterWithProgress.getCount()).thenReturn(count);
        return this;
    }

    StateBuilder counterWithProgressProgressValue(int value) {
        when(mockCounterWithProgress.getProgress()).thenReturn(value);
        return this;
    }

    ActivityTestRule<CounterActivity>  createRule(){

        return new ActivityTestRule<CounterActivity>(CounterActivity.class) {
            @Override
            protected void beforeActivityLaunched() {

                //get hold of the application
                App app = (App) InstrumentationRegistry.getTargetContext().getApplicationContext();
                OG.setApplication(app, WorkMode.SYNCHRONOUS);

                //inject our mocks so our UI layer will pick them up
                OG.putMock(CounterWithLambdas.class, mockCounterWithLambdas);
                OG.putMock(CounterWithProgress.class, mockCounterWithProgress);
            }

        };
    }

}

File: example-kt-01reactiveui/src/androidTest/java/foo/bar/example/forereactiveui/ui/wallet/StateBuilder.kt
package foo.bar.example.forereactiveui.ui.wallet

import androidx.test.core.app.ApplicationProvider
import androidx.test.rule.ActivityTestRule
import co.early.fore.kt.core.delegate.ForeDelegateHolder
import co.early.fore.kt.core.delegate.TestDelegateDefault
import foo.bar.example.forereactiveui.App
import foo.bar.example.forereactiveui.OG
import foo.bar.example.forereactiveui.feature.wallet.Wallet
import io.mockk.every

/**
 *
 */
class StateBuilder internal constructor(private val mockWallet: Wallet) {

    internal fun withMobileWalletMaximum(totalFundsAvailable: Int): StateBuilder {

        every {
            mockWallet.mobileWalletAmount
        } returns totalFundsAvailable

        every {
            mockWallet.savingsWalletAmount
        } returns 0

        every {
            mockWallet.canDecrease()
        } returns true

        every {
            mockWallet.canIncrease()
        } returns false

        return this
    }

    internal fun withMobileWalletHalfFull(savingsWalletAmount: Int, mobileWalletAmount: Int): StateBuilder {

        every {
            mockWallet.mobileWalletAmount
        } returns mobileWalletAmount

        every {
            mockWallet.savingsWalletAmount
        } returns savingsWalletAmount

        every {
            mockWallet.canDecrease()
        } returns true

        every {
            mockWallet.canIncrease()
        } returns true

        return this
    }

    internal fun withMobileWalletEmpty(totalFundsAvailable: Int): StateBuilder {

        every {
            mockWallet.mobileWalletAmount
        } returns 0

        every {
            mockWallet.savingsWalletAmount
        } returns totalFundsAvailable

        every {
            mockWallet.canDecrease()
        } returns false

        every {
            mockWallet.canIncrease()
        } returns true

        return this
    }

    fun createRule(): ActivityTestRule<WalletsActivity> {
        return object : ActivityTestRule<WalletsActivity>(WalletsActivity::class.java) {
            override fun beforeActivityLaunched() {

                //get hold of the application
                val app: App = ApplicationProvider.getApplicationContext<App>() as App
                OG.setApplication(app)

                ForeDelegateHolder.setDelegate(TestDelegateDefault())

                //inject our mocks so our UI layer will pick them up
                OG.putMock(Wallet::class.java, mockWallet)
            }
        }
    }
}
